File: archivo-historico/ejemplos/nerve/07-bot-discord-nerve/ejemplo.go
/**
 * Ejemplo: Bot básico con discordgo
 * 
 * Qué enseña este ejemplo:
 *   - Cómo configurar e iniciar una sesión de discordgo en Go.
 *   - Configurar los "Intents" para especificar qué eventos quieres.
 *   - Añadir un "handler" (manejador) para eventos de nuevos mensajes.
 * 
 * Para tu reto (Bot Discord Nerve):
 *   En vez de solo responder con "¡Hola!", conectarás el bot de Discord con
 *   tu cliente Nerve. Cuando llegue un mensaje a Discord, usarás tu cliente 
 *   Nerve para notificar a la red, y viceversa.
 * 
 * Glosario:
 *   discordgo.New()         — Crea una nueva sesión del bot.
 *   dg.Identify.Intents     — Define qué eventos de Discord recibirá el bot (ej. GuildMessages).
 *   dg.AddHandler()         — Registra una función que se llamará cuando ocurra un evento.
 *   s.ChannelMessageSend()  — Envía un mensaje a un canal de Discord.
 */

package main

import (
	"fmt"
	"os"
	"os/signal"
	"strings"
	"syscall"

	"github.com/bwmarrin/discordgo"
)

func main() {
	// Reemplaza esto con tu token real o usa variables de entorno
	token := os.Getenv("DISCORD_TOKEN")
	if token == "" {
		token = "TU_TOKEN_AQUI"
	}

	if token == "TU_TOKEN_AQUI" {
		fmt.Println("⚠️ ADVERTENCIA: Reemplaza 'TU_TOKEN_AQUI' con el token de tu bot de Discord.")
		fmt.Println("Puedes obtenerlo en: https://discord.com/developers/applications")
		return
	}

	// 1. Crear una nueva sesión de Discord usando el token del bot
	dg, err := discordgo.New("Bot " + token)
	if err != nil {
		fmt.Println("Error creando la sesión de Discord:", err)
		return
	}

	// 2. Registrar el manejador (handler) para el evento MessageCreate
	dg.AddHandler(messageCreate)

	// 3. Configurar los Intents (permisos). Necesitamos los de mensajes.
	// Nota: Para MessageContent, asegúrate de habilitarlo en el portal de desarrolladores.
	dg.Identify.Intents = discordgo.IntentsGuildMessages | discordgo.IntentsMessageContent

	// 4. Abrir una conexión con Discord
	err = dg.Open()
	if err != nil {
		fmt.Println("Error abriendo conexión:", err)
		return
	}

	// Esperar hasta que se reciba una señal para terminar (CTRL+C)
	fmt.Println("El bot está funcionando. Presiona CTRL-C para salir.")
	sc := make(chan os.Signal, 1)
	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM, os.Interrupt)
	<-sc

	// 5. Cerrar la sesión de forma limpia
	dg.Close()
}

// messageCreate se llamará CADA VEZ que se envíe un mensaje a un canal al que el bot tenga acceso
func messageCreate(s *discordgo.Session, m *discordgo.MessageCreate) {
	// Ignorar todos los mensajes creados por el bot mismo
	if m.Author.ID == s.State.User.ID {
		return
	}

	// Si el mensaje dice "ping", responder con "pong"
	if strings.ToLower(m.Content) == "ping" {
		fmt.Printf("[%s] dijo ping, respondiendo...\n", m.Author.Username)
		
		// Enviar el mensaje al mismo canal donde se recibió el ping
		s.ChannelMessageSend(m.ChannelID, "pong 🏓")
	}
}

File: archivo-historico/ejemplos/nerve/07-bot-discord-nerve/ejemplo.js
/**
 * Ejemplo: Bot básico con discord.js
 * 
 * Qué enseña este ejemplo:
 *   - Cómo configurar e iniciar un cliente de discord.js.
 *   - El uso de "GatewayIntents" para especificar qué eventos quieres.
 *   - Manejo de eventos con `.on` y `.once`.
 * 
 * Para tu reto (Bot Discord Nerve):
 *   En vez de solo responder con "¡Hola!", conectarás el bot de Discord con
 *   tu cliente Nerve. Cuando llegue un mensaje a Discord, usarás tu cliente 
 *   Nerve para notificar a la red, y viceversa.
 * 
 * Glosario:
 *   Client()                 — Instancia principal del bot de Discord.
 *   GatewayIntentBits        — Sistema de permisos de Discord. `GuildMessages` lee eventos de mensajes,
 *                              y `MessageContent` es necesario para leer el texto en sí.
 *   client.once('ready')     — Se ejecuta UNA SOLA VEZ cuando el bot termina de conectarse.
 *   client.on('messageCreate') — Se ejecuta CADA VEZ que se envía un mensaje.
 *   message.reply()          — Envía un mensaje respondiendo (etiquetando) al usuario original.
 */

const { Client, GatewayIntentBits } = require('discord.js');

// 1. Configurar el cliente y los Intents
const client = new Client({ 
    intents: [
        GatewayIntentBits.Guilds,           // Para poder estar en servidores
        GatewayIntentBits.GuildMessages,    // Para recibir eventos de mensajes
        GatewayIntentBits.MessageContent    // Para poder leer el contenido de los mensajes
    ] 
});

// 2. Evento 'ready': Cuando el bot se conecta exitosamente
client.once('ready', () => {
    console.log(`✅ Conectado exitosamente como ${client.user.tag}`);
});

// 3. Evento 'messageCreate': Cuando alguien envía un mensaje
client.on('messageCreate', async (message) => {
    // Ignorar mensajes enviados por bots (incluyendo el nuestro)
    if (message.author.bot) return;

    // Si alguien dice "ping", respondemos "pong"
    if (message.content.toLowerCase() === 'ping') {
        console.log(`[${message.author.tag}] dijo ping, respondiendo...`);
        // Usar await porque enviar un mensaje es una operación de red
        await message.reply('pong 🏓');
    }
});

// 4. Iniciar el bot
function main() {
    // Debes reemplazar esto con tu token real o usar variables de entorno
    const token = process.env.DISCORD_TOKEN || "TU_TOKEN_AQUI";

    if (token === "TU_TOKEN_AQUI") {
        console.log("⚠️ ADVERTENCIA: Reemplaza 'TU_TOKEN_AQUI' con el token de tu bot de Discord.");
        console.log("Puedes obtenerlo en: https://discord.com/developers/applications");
        return;
    }

    console.log("Iniciando bot...");
    client.login(token);
}

main();
